packages/console/src/pages/GetStarted/index.tsx

```diff
@@ -1,7 +1,126 @@
+   * e.g. React, Angular, Vue, etc. for application. Or Github, WeChat, etc. for connector.
+   */
+  defaultSubtype?: string;
+  bannerComponent?: ReactNode;
+  onClose?: () => void;
+  onComplete?: () => void;
+  onToggleSteps?: () => void;
+}>;
+
+const onClickFetchSampleProject = (projectName: string) => {
+  const sampleUrl = `https://github.com/logto-io/js/tree/master/packages/${projectName}-sample`;
+  window.open(sampleUrl, '_blank');
+};
+
+const GetStarted = ({
+  title,
+  subtitle,
+  type,
+  defaultSubtype,
+  bannerComponent,
+  onClose,
+  onComplete,
+  onToggleSteps,
+}: Props) => {
+  const [subtype, setSubtype] = useState(defaultSubtype);
+  const [activeStepIndex, setActiveStepIndex] = useState<number>(-1);
+  const steps = useGetStartedSteps(type, subtype) ?? [];
+
+  const stepReferences = useMemo(
+    () => Array.from({ length: steps.length }).map(() => createRef<HTMLDivElement>()),
+    [steps.length]
+  );
+
+  useEffect(() => {
+    if (activeStepIndex > -1) {
+      const activeStepReference = stepReferences[activeStepIndex];
+      activeStepReference?.current?.scrollIntoView({ block: 'start', behavior: 'smooth' });
+    }
+  }, [activeStepIndex, stepReferences]);
+
+  return (
+    <div className={styles.container}>
+      <div className={styles.header}>
+        <IconButton size="large" onClick={onClose}>
+          <Close />
+        </IconButton>
+        <div className={styles.separator} />
+        <CardTitle size="small" title={<DangerousRaw>{title}</DangerousRaw>} subtitle={subtitle} />
+        <Spacer />
+        <Button type="plain" size="small" title="general.skip" onClick={onClose} />
+        {subtype && (
+          <Button
+            type="outline"
+            title="admin_console.applications.get_started.get_sample_file"
+            onClick={() => {
+              onClickFetchSampleProject(subtype);
+            }}
+          />
+        )}
+      </div>
+      <div className={styles.content}>
+        {isValidElement(bannerComponent) &&
+          cloneElement(bannerComponent, {
+            className: styles.banner,
+            onChange: setSubtype,
+            onToggle: () => {
+              setActiveStepIndex(0);
+            },
+          })}
+        {steps.map((step, index) => {
+          const isFinalStep = index === steps.length - 1;
+
+          return (
+            <Step
+              key={step.title}
+              ref={stepReferences[index]}
+              data={step}
+              index={index}
+              isCompleted={activeStepIndex > index}
+              isExpanded={activeStepIndex === index}
+              isFinalStep={isFinalStep}
+              onComplete={onComplete}
+              onNext={() => {
+                setActiveStepIndex(index + 1);
+              }}
+              onToggle={() => {
+                setActiveStepIndex(index);
+                onToggleSteps?.();
+              }}
+            />
+          );
+        })}
```

Comment:
extract to a use memo?

Reply:
We just talked about it and we don't think a useMemo would help in this case. Resolved.